### PR TITLE
MM-50087_Add marketplace button to apps bar

### PR DESCRIPTION
**Pull request by julmondragon**
#### Summary
Show the marketplace button in the App Bar.
Additionally standardized for all places where the marketplace can be opened:
- Marketplace should be enabled and user should have `SYSCONSOLE_WRITE_PLUGINS` permission.
- Label should read "App Marketplace"
- Icon should be `view-grid-plus-outline` Screen Shot 2023-03-06 at 11 52 58 AM
- Current event tracking for Marketplace modal opened is now enhanced to report the source of the action.
  ```
    properties: {
        category: "plugins",
        type: "ui_marketplace_opened",
        from: "actions_menu" | "app_bar" | "channel_header" | "command" | "open_plugin_install_post" | "product_menu"
        ...
    }
  ```

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-50087

##### App Bar enabled
Mar-14-2023 14-03-45

#### Release Note
```release-note
NONE
```

### CODE REVIEW

File: plugins/channel_header_plug/channel_header_plug.tsx
Comment (pvev) [inline plugins/channel_header_plug/channel_header_plug.tsx:263]: nit: place these places in global constants and avoid magic strings

**Review by pvev — COMMENTED** [review]
LGTM! thanks @julmondragon 🚀

**Review by pvev — APPROVED** [review]

File: components/app_bar/app_bar.tsx
Comment (julmondragon) [reply components/app_bar/app_bar.tsx]: Updated

**Comment by stevemudie** [comment]
This works great @julmondragon!

I'll hold off approving it for now though in case changes come along that need to be retested.

**Review by JulienTant — CHANGES_REQUESTED** [review]
This is kindof a light request change because most of everything is LGTM ^^ I believe checking only for sysadmin to show to button is not the best way to handle this, there's a specific permission that should be checked that normal users could have.

**Comment by esethna** [comment]
Thanks @julmondragon! Let's hold this on @annelieseklein's review. There are a few things that look a bit off, such as the pipe divider under the boards and playbooks buttons, and the hover effect not matching our design guidelines

**Review by esethna — APPROVED** [review]

File: components/app_bar/app_bar.tsx
Comment (julmondragon) [reply components/app_bar/app_bar.tsx:80]: Thanks for the input!
Requirements on the ticket specified only system admins, and reviewing other places where marketplace is accessed are checking `SYSCONSOLE_READ_PLUGINS`.
At least I think we need to make sure it's enabled on the server, will work on adding it.

And maybe @esethna can confirm if should include non system admin users with either `SYSCONSOLE READ_PLUGINS` or `SYSCONSOLE_WRITE_PLUGINS`.

Comment (JulienTant) [reply components/app_bar/app_bar.tsx:80]: Year FIY I took this info from: https://github.com/mattermost/mattermost-webapp/blob/master/components/global_header/left_controls/product_menu/product_menu_list/product_menu_list.tsx#L207-L224

This is the condition we use to show or not the marketplace in the "product" menu

Comment (marianunez) [reply components/app_bar/app_bar.tsx:80]: +1 - we should use the permissions systems given that if there are custom permission schemes, they may grant plugin write access to users that may not be admins. Admins have the `Permissions.SYSCONSOLE_WRITE_PLUGINS]` so that should be enough. 

As Julian pointed out, the action menu does just check for `isAdmin` to show or not that marketplace entry. We should fix that. Are you open to include that as part of this PR?

For now we don't allow to open the marketplace if they have only `SYSCONSOLE_READ_PLUGINS` access or regular users. This is something we could introduce in the future with the functionality of Notify Admin to install plugins. cc/ @esethna

File: components/app_bar/app_bar.test.tsx
Comment (marianunez) [inline components/app_bar/app_bar.test.tsx:146]: Could we add the negative test that doesn't show for non-system admin?

File: components/plugin_marketplace/marketplace_modal.test.tsx
Comment (marianunez) [inline components/plugin_marketplace/marketplace_modal.test.tsx:210]: 💯

**Review by marianunez — APPROVED** [review]
Thanks @julmondragon! Approving pending the permissions change discussed above

File: plugins/channel_header_plug/channel_header_plug.tsx
Comment (hanzei) [inline plugins/channel_header_plug/channel_header_plug.tsx:283]: Why does the tooltip say `App Marketplace`? The Marketplace currently only contains a single app and the rest are plugins.

**Review by catalintomai — APPROVED** [review]

File: components/app_bar/app_bar.tsx
Comment (julmondragon) [reply components/app_bar/app_bar.tsx:80]: @marianunez Sure, will add the privilege check in the action menu.
I see that the command `/marketplace` is checking for the read plugins privilege (`SYSCONSOLE_READ_PLUGINS`) https://github.com/mattermost/mattermost-webapp/blob/master/actions/command.ts#L131-L133
Should this be changed as well to check the write one instead?

File: plugins/channel_header_plug/channel_header_plug.tsx
Comment (julmondragon) [reply plugins/channel_header_plug/channel_header_plug.tsx:283]: @hanzei  As specified in the ticket, this should follow the current implementation in the post hover menu: Message actions > Apps Marketplace
Screen Shot 2023-03-06 at 11 44 13 AM

Should texts be changed in both instances to just 'Marketplace'?
BTW it reads this way in the product menu:
Screen Shot 2023-03-06 at 11 49 40 AM

Last thing, noticed that we'll have a difference in the icon used in the marketplace item. Should we also make all the instances to match? 
- Screen Shot 2023-03-06 at 11 52 58 AM `view-grid-plus-outline`
- Screen Shot 2023-03-06 at 11 53 12 AM `apps`

cc @esethna @marianunez @annelieseklein

File: components/app_bar/app_bar.tsx
Comment (julmondragon) [reply components/app_bar/app_bar.tsx]: @marianunez Updated to check for both Marketplace enabled and `SYSCONSOLE_WRITE_PLUGINS` permission in:
- Channel header
- App bar
- Post actions menu

Pending confirmation on the `/marketplace` command.

**Comment by julmondragon** [comment]
> Thanks @julmondragon! Let's hold this on @annelieseklein's review. There are a few things that look a bit off, such as the pipe divider under the boards and playbooks buttons (when there's no apps below it), and the hover effect not matching our design guidelines
> 
> Functionally it looks good though so approving from the PM side

@esethna Already solved the pipe divider. Waiting on @annelieseklein for design of button's states.

File: components/app_bar/app_bar.test.tsx
Comment (julmondragon) [reply components/app_bar/app_bar.test.tsx]: Added

File: plugins/channel_header_plug/channel_header_plug.tsx
Comment (esethna) [reply plugins/channel_header_plug/channel_header_plug.tsx:283]: Here's my proposal:

- Opening the marketplace should use the `view-grid-plus-outline`, since users are opening the marketplace to add new apps/plugins. Propose we change that icon in the header product menu so we have consistency with the post hover menu and apps bar icon. 
- We should call it the "App Marketplace". Even though it's technically mostly plugins, the terminology of "Apps" helps users understand what the marketplace is for. 

cc// @asatkinson @julmondragon

**Comment by annelieseklein** [comment]
@julmondragon i'm having trouble accessing the most recent server for some reason - ill look at an older server

**Comment by julmondragon** [comment]
> @julmondragon i'm having trouble accessing the most recent server for some reason....any idea why?

@annelieseklein It looks like a general issue with spinwicks for PRs at the moment. Please let me know if in the meantime there is something I can provide (gif, video, ...), or if there is any figma design I can use to style the button.

**Comment by annelieseklein** [comment]
> > @julmondragon i'm having trouble accessing the most recent server for some reason....any idea why?
> 
> @annelieseklein It looks like a general issue with spinwicks for PRs at the moment. Please let me know if in the meantime there is something I can provide (gif, video, ...), or if there is any figma design I can use to style the button.

ofc! heres the figma design guidelines for the button component - pease let me know if you have any questions https://www.figma.com/file/2uYWxjnMJ9IQDOba9b9bfV/Components---Button?node-id=3558%3A13159&t=J6nPOAt0O8CY07P2-1

**Comment by julmondragon** [comment]
PR updated with feedback, design guidelines and the spinwick is available now.
@annelieseklein @marianunez @esethna @stevemudie

**Comment by esethna** [comment]
Thanks! @annelieseklein is that the intended click effect? It seems different than some other buttons in the app (ie channel info, saved posts or recent mentions)

**Comment by annelieseklein** [comment]
@esethna are you talking about the click effect for the app bar button? if so, yes - here is the design file for the states for the app bar specifically https://www.figma.com/file/43xvK7GOEPh6VJ31UfRvSg/MM-28021-Channel-Info-RHS?node-id=53%3A79124&t=H9KKtXDH6YI0RTt6-1

**Comment by julmondragon** [comment]
@JulienTant @stevemudie This is ready for final review.

**Review by JulienTant — APPROVED** [review]
LGTM @julmondragon

**Comment by esethna** [comment]
@annelieseklein sorry it feels off as it's not consistent with similar buttons of that pattern (My fault for rushing the ticket without a proper design).

If we are going with a square hover effect (which I think works better for this icon since it's not round like the other buttons in the app bar) then I think we should be consistent with similar buttons that follow that pattern and not do the bright border on click.

Apps marketplace button:
image

Message priority button: 
image

**Comment by annelieseklein** [comment]
@esethna so that is different than the apps button -  I wasn't sure what you were referring to.

@julmondragon this is the icon button that I mentioned to you that is in this file: https://www.figma.com/file/n4JufqTFIQYKwmTERUnBqZ/Components---Icon-Buttons?node-id=9%3A995&t=errOhtguZkdAtAPF-1

Screen Shot 2023-03-14 at 11 06 24 AM

**Comment by julmondragon** [comment]
@annelieseklein @esethna Updated style for button active.

**Review by annelieseklein — APPROVED** [review]
LGTM!

**Review by stevemudie — APPROVED** [review]

**Comment by julmondragon** [comment]
Replaced with https://github.com/mattermost/mattermost-server/pull/22653

**Pull request closed without merging**
